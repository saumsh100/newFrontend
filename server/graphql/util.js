
import { getOffsetWithDefault, offsetToCursor } from 'graphql-relay';

/**
 * A simple function that accepts an array and connection arguments, and returns
 * a connection object for use in GraphQL
 *
 * This is a fork of connectionFromArraySlice
 * from https://github.com/graphql/graphql-relay-js/blob/master/src/connection/arrayconnection.js#L65
 *
 * Diferente form the original, is this case we build the cursor and pageInfo based on args
 * whitout slicing the array.
 * That way we can pass the already sliced or filtered array to the function.
 *
 * For now, this only works with forward pagination.
 */
// eslint-disable-next-line import/prefer-default-export
export function connectionFromArrayWithoutSlice(array, args, meta) {
  const { after, first, last } = args;
  const { arrayLength } = meta;

  if (typeof first === 'number') {
    if (first < 0) {
      throw new Error('Argument "first" must be a non-negative integer');
    }
  }
  if (typeof last === 'number') {
    if (last < 0) {
      throw new Error('Argument "last" must be a non-negative integer');
    }
  }

  const afterOffset = getOffsetWithDefault(after, 0);

  const startOffset = Math.max(afterOffset, -1) + 1;
  const endOffset = startOffset + first;

  const edges = array.map((value, index) => ({
    cursor: offsetToCursor(startOffset + index),
    node: value,
  }));

  const firstEdge = edges[0];
  const lastEdge = edges[edges.length - 1];
  const lowerBound = after ? afterOffset + 1 : 0;
  const upperBound = arrayLength;

  return {
    edges,
    pageInfo: {
      startCursor: firstEdge ? firstEdge.cursor : null,
      endCursor: lastEdge ? lastEdge.cursor : null,
      hasPreviousPage: typeof last === 'number' ? startOffset > lowerBound : false,
      hasNextPage: typeof first === 'number' ? endOffset < upperBound : false,
    },
  };
}
