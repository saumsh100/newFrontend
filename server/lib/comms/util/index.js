
import groupBy from 'lodash/groupBy';
import map from 'lodash/map';
import toArray from 'lodash/toArray';

const ATTRS = {
  email: 'email',
  sms: 'mobilePhoneNumber',
  phone: 'mobilePhoneNumber',
};

const ERROR_CODES = {
  email: '1100',
  sms: '1200',
  phone: '1300',
};

const PREFERENCE_TYPES = {
  email: 'emailNotifications',
  sms: 'sms',
  phone: 'phone',
};

const NO_PREFERENCE_ERROR_CODES = {
  email: '2100',
  sms: '2200',
  phone: '2300',
};

/**
 * cannotSend
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function cannotSend(patient, primaryType) {
  // If it is undefined return the error code
  if (!patient[ATTRS[primaryType]]) {
    return ERROR_CODES[primaryType];
  }
}

/**
 * noPreference
 *
 * @param patient
 * @param primaryType
 * @returns {{errors: Array, success: Array}}
 */
export function noPreference(patient, primaryType) {
  const preferences = patient.preferences || {};

  // If it is false return errorCode
  if (!preferences[PREFERENCE_TYPES[primaryType]]) {
    return NO_PREFERENCE_ERROR_CODES[primaryType];
  }
}

/**
 * generateOrganizedPatients will accept an array of patients and an
 * array of primaryTypes and will return the errors and successes based
 * on patientData. The point is to organize the failed responses so that
 * certain jobs can save a bulk amount to save time
 *
 * @param patients
 * @param primaryTypes (ie.// ['email', 'sms'])
 * @returns {{errors: Array, success: Array}}
 */
export function generateOrganizedPatients(patients, primaryTypes) {
  const errors = [];
  const success = [];

  for (const patient of patients) {
    primaryTypes.forEach((primaryType) => {
      const errorCode = cannotSend(patient, primaryType) || noPreference(patient, primaryType);
      if (errorCode) {
        errors.push({
          patient,
          errorCode,
          primaryType,
        });
      } else {
        success.push({ primaryType, patient });
      }
    });
  }

  return { errors, success };
}

/**
 * organizeForOutbox is a function that accepts an array and a selectorPredicate
 * and will group the array based on the selectorPredicate, and then map over that array
 * to return outbox items that have primaryTypes grouped together.
 *
 * NOTE: this is primarily a layer ontop of the mapPatientsTo____ calls that will take the flat
 * array with primaryType and bring it back into primaryTypes
 *
 * @param array (ie.// [{ patient, appointment, primaryType: 'sms' }, { patient, appointment, primaryType: 'email' }])
 * @param selectorPredicate (ie.// ({ patient, appointment }) => `${patient.id}_${appointment.id}`)
 * @param mergePredicate
 * @returns [organizedArray] (ie.// [{ patient, appointment, primaryTypes: ['sms', 'email'] }])
 */
export function organizeForOutbox(array, selectorPredicate, mergePredicate) {
  const groupedMap = groupBy(array, selectorPredicate);
  return toArray(map(groupedMap, mergePredicate));
}
