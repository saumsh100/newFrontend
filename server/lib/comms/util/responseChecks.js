
import find from 'lodash/find';

const THE_MAGIC_WORDS = {
  C: true,
  Y: true,
};

/**
 * getIndicesOf will return the indices in the text
 * of all the magicWord
 *
 * @param text
 * @param magicWord
 * @returns {Array}
 */
export function getIndicesOf(text, magicWord) {
  const magicWordLen = magicWord.length;
  if (magicWordLen === 0) {
    return [];
  }

  const indices = [];
  const regexp = new RegExp(magicWord, 'g');

  let match;
  while (match = regexp.exec(text))  {
    indices.push(match.index);
  }

  return indices;
}

/**
 *
 * @param char
 */
export function isWordy(char) {
  return /[a-zA-Z0-9'"]/.test(char);
}

/**
 *
 * @param text
 * @param magicWord
 */
export function isNotInWord(text, magicWord) {
  // If exactly equal to magicWord then don't go any further
  if (text === magicWord) {
    return true;
  }

  // Get all indicies of the magicWord in the text
  const indices = getIndicesOf(text, magicWord);

  // Loop through every occurance and see if any of the before character or
  // after character is alpha-numeric. If so, then return false
  for (const index of indices) {
    const beforeCharacter = text[index - 1];
    const afterCharacter = text[index + 1];
    if (!beforeCharacter && !isWordy(afterCharacter)) {
      return true;
    }

    if (!afterCharacter && !isWordy(beforeCharacter)) {
      return true;
    }

    if (beforeCharacter && !isWordy(beforeCharacter) && afterCharacter && !isWordy(afterCharacter)) {
      return true;
    }
  }

  return false;
}

/**
 * Determines if a certain string of text is a valid confirmation response
 *
 * @param text
 * @param magicWord
 * @returns {boolean}
 */
export function wordCheck(text, magicWord) {
  // If exactly equal to magicWord then don't go any further
  if (text === magicWord) {
    return true;
  }

  // If the text doesn't have the magic word at all, don't bother continuing!
  const hasMagicWord = text.includes(magicWord);

  // If the magicWord is part of another word, it doesn't count
  // ex.// magicWord='C'  text='Can't make it'
  const magicWordIsNotInWord = isNotInWord(text, magicWord);

  // This function plays nicer with checking AND for all booleans
  return hasMagicWord && magicWordIsNotInWord;
}

/**
 * isSmsConfirmationResponse determines if a certain string of text is
 * a valid confirmation response
 *
 * @param text
 * @param magicWords
 * @returns {boolean}
 */
export function isSmsConfirmationResponse(text, magicWords = THE_MAGIC_WORDS) {
  text = text.toUpperCase();

  return typeof magicWords !== 'string'
    ? !!find(magicWords, (value, key) => wordCheck(text, key))
    : wordCheck(text, magicWords);
}
