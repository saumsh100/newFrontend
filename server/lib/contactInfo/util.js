
const orderBy = require('lodash/orderBy');

/**
 * isHead checks if a given patient is the head of the given family by comparing accountId,
 * familyId and headId.
 * @param family
 * @param patient
 * @return {boolean}
 */
const isHead = (family, patient) =>
  !!family &&
  !!patient &&
  family.accountId === patient.accountId &&
  family.id === patient.familyId &&
  family.headId === patient.id;

/**
 * isActive checks if a patient has status equals to the string 'Active'
 * @param patient
 * @return {boolean}
 */
const isActive = ({ status }) => status === 'Active';

/**
 * getFromNewestCreatedFamily orders the a list of patients by its family creation date and
 * grabs the first.
 * @param patients
 */
const getFromNewestCreatedFamily = patients =>
  orderBy(
    patients,
    ({ family: { pmsCreatedAt, createdAt } }) => pmsCreatedAt || createdAt,
    'desc',
  )[0];

/**
 * getOldest orders a list of patients by their birth dates and grabs the first, thus the oldest.
 * @param patients
 */
const getOldest = patients => orderBy(patients, p => p.birthDate, 'asc')[0];

/**
 * getPatientAge is a utility function to extract the patient age form its birth date by year
 * diffing. This was created using plain JS so we can reduce the usage of moment.js.
 * @param patient
 * @return {number}
 */
const getPatientAge = (patient) => {
  const ageDate = new Date(Date.now() - new Date(patient.birthDate).getTime());
  return Math.abs(ageDate.getUTCFullYear() - 1970);
};

/**
 * adultChecker is a configurable function builder that creates a function that will check if an age
 * of a patient is within the adult range provided to it.
 * @param min
 * @param max
 * @return {function(*=): boolean}
 */
const adultChecker = (min, max) => patient =>
  min < getPatientAge(patient) && getPatientAge(patient) < max;

/**
 * grabPoC is a function that takes a patient list and returns a function that filters this list by
 * the given predicate, in case of the filter returns more than one result the fallbackHandler is
 * applied to the list.
 * This function use partially applied arguments so we can reuse the same list of patients but with
 * different predicates and fallbackHandlers.
 * @param patients
 * @return {Function}
 */
const grabPoC = patients => (filterPredicate, fallbackHandler) => {
  const filteredPatients = patients.filter(filterPredicate);
  if (filteredPatients.length === 1) {
    return filteredPatients[0];
  } else if (filteredPatients.length > 1) {
    return fallbackHandler(filteredPatients);
  }

  return undefined;
};

module.exports = {
  adultChecker,
  grabPoC,
  isHead,
  isActive,
  getFromNewestCreatedFamily,
  getOldest,
};
