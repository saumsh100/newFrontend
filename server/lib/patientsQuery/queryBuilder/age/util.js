
import { setDateToTimezone } from '@carecru/isomorphic';

/**
 * Compose a query-builder-ready parameters
 * @param params
 * @return {*}
 */
export function prepareQueryParams(params) {
  if (Array.isArray(params)) {
    return calculateDatesArray(params);
  }

  const reformattedParameters = Object.entries(params).map(([key, value]) => {
    const formattedValue = Array.isArray(value)
      ? calculateDatesArray(value)
      : calculateDate(value);
    return { [key]: formattedValue };
  });

  return Object.assign(...reformattedParameters);
}

/**
 * Calculate the date difference for an array of years.
 * @param {Array} yearValue
 * @return {string[]}
 */
function calculateDatesArray(yearValue) {
  return yearValue.map(calculateDate).sort();
}

/**
 * Calculate a difference in years, by subtracting the year given as parameters from the
 * current date.
 * @param years
 * @return {string}
 */
export function calculateDate(years) {
  const now = setDateToTimezone()
    .hours(0)
    .minutes(0)
    .seconds(0)
    .milliseconds(0);

  return now.subtract(years, 'years').toISOString();
}
