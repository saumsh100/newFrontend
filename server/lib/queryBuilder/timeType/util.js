
import { time } from '@carecru/isomorphic';

/**
 * Normalize a whole request object to have a correct date.
 * @param requestObject
 * @return {*}
 */
const normalizeRequest = (requestObject) => {
  if (typeof requestObject === 'string') {
    return normalizeTime(requestObject);
  }
  if (Array.isArray(requestObject)) {
    return requestObject.map(normalizeTime);
  }

  const normalizedObjects = Object.entries(requestObject)
    .map(([key, value]) => ({ [key]: normalizeRequest(value) }));

  return Object.assign(...normalizedObjects);
};

/**
 * Create a complete date-time from the requested time value.
 * @param requestTime
 * @return {string}
 */
const normalizeTime = (requestTime) => {
  const [hours, minutes] = requestTime.split(':');
  return time(hours, minutes).toISOString();
};


export default normalizeRequest;
