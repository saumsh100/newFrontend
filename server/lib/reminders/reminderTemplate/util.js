
import { sortAsc, setDateToTimezone } from '@carecru/isomorphic';
import { getReminderType } from '../createReminderText';
import isFeatureFlagEnabled from '../../featureFlag';

export function generateFamilyDetails(dependants, timezone) {
  if (dependants.length > 1) {
    const familyDetails = dependants
      .sort(
        (
          { appointment: { startDate: a } },
          { appointment: { startDate: b } },
        ) => sortAsc(a, b),
      )
      .map(({ appointment: { startDate }, firstName, lastName }) => {
        const { date, time } = getDateAndTime(startDate, timezone);
        return `\n${firstName} ${lastName}\n${date} at ${time}\n`;
      })
      .join('');

    return { familyDetails };
  }

  const [familyMember] = dependants;
  const {
    appointment: { startDate },
  } = familyMember;

  return {
    familyMember,
    ...getDateAndTime(startDate, timezone),
  };
}

export function getDateAndTime(date, timezone) {
  const mDate = setDateToTimezone(date, timezone);
  return {
    date: mDate.format('MMMM Do'),
    time: mDate.format('h:mma'),
  };
}

/**
 * Feature flag is here so that both the Preview API and the actual
 * reminders job can utilize the same function
 * @param account
 * @return {Promise<boolean>}
 */
export async function shouldUseLegacyTemplate(account) {
  const isCustomizable = isFeatureFlagEnabled(
    'customizable-reminders-templates',
    false,
    {
      userId: 'carecru-api',
      accountId: account.id,
      enterpriseId: account.enterpriseId,
    },
  );

  return !isCustomizable;
}

export function getReminderTemplateType({
  account,
  appointment,
  reminder,
  currentDate,
  dependants,
}) {
  if (isFamily(dependants)) {
    if (appointment) {
      return 'self';
    }

    return dependants.length === 1 ? 'single' : 'multiple';
  }

  const templateTypes = {
    weeksAway: 'weeks-away',
    weekAway: 'week-away',
    sameWeek: 'same-week',
    sameDay: 'same-day',
  };

  // Don't pass in appointment so it uses the reminder.interval instead
  const reminderType = getReminderType({
    account,
    appointment,
    reminder,
    currentDate,
  });

  return templateTypes[reminderType];
}

export function isFamily(dependants) {
  return dependants && dependants.length > 0;
}
