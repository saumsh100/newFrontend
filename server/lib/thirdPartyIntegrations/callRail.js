
import axios from 'axios';
import { statusError } from '@carecru/isomorphic';
import { callrails, fullHostUrl } from '../../config/globals';

const CALL_RAIL_URL = `https://api.callrail.com/v2/a/${callrails.apiAccount}`;
const errorHandling = ({ customMessage, e }) => {
  console.error(customMessage);
  console.error(e);
  throw statusError(e.status, e.data ? e.data.error : e.statusText);
};

function generateCallRailRequestHeaders() {
  return { Authorization: `Token token=${callrails.apiKey}` };
}

function checkForIdExistence(id) {
  if (!id) {
    throw statusError(statusError.NOT_FOUND, 'Account does not have callRail account associated with');
  }
}

async function makeCallRailRequest({ method, url, data, params }) {
  return axios({
    method,
    url,
    headers: generateCallRailRequestHeaders(),
    data,
    params,
    json: true,
  });
}

/**
 * Update the CallRail account company name for a practice account.
 * @throws StatusError
 * @param account
 * @returns {Promise<void>}
 */
export async function updateCompanyName({ name, callrailId }) {
  checkForIdExistence(callrailId);
  try {
    const { data } = await makeCallRailRequest({
      method: 'PUT',
      url: `${CALL_RAIL_URL}/companies/${callrailId}.json`,
      data: { name },
    });

    return data;
  } catch (e) {
    errorHandling({
      customMessage: 'Failed to update the company name for the account',
      e,
    });
  }
}

/**
 * Update the CallRail account phone number for a practice account.
 * The old and the new phone number are required because for one practice,
 * multiple trackers may exist.
 * @throws StatusError
 * @param account
 * @param oldPhoneNumber
 * @param newPhoneNumber
 * @returns {Promise<*>}
 */
export async function updatePhoneNumber(account, oldPhoneNumber, newPhoneNumber) {
  const { callrailId } = account;
  checkForIdExistence(callrailId);
  if (!oldPhoneNumber || !newPhoneNumber) {
    throw statusError(statusError.BAD_REQUEST, 'Please provide both the oldPhoneNumber and the newPhoneNumber');
  }

  try {
    const activeTrackers = await getActiveTrackers(callrailId);
    await Promise.all(activeTrackers
      .filter(({ destination_number }) => destination_number === oldPhoneNumber)
      .map(({ id }) => makeCallRailRequest({
        method: 'PUT',
        url: `${CALL_RAIL_URL}/trackers/${id}.json`,
        data: {
          call_flow: {
            type: 'basic',
            recording_enabled: true,
            destination_number: newPhoneNumber,
          },
        },
      })));

    return exports.getCallRailInformation(account);
  } catch (e) {
    errorHandling({
      customMessage: 'Failed to update the tracking phone number for the account',
      e,
    });
  }
}

/**
 * Retrieve the CallRail information for a practice account.
 * The response will be an JSON object consist of three part: company, trackers, webhooks
 * @throws StatusError
 * @param callrailId
 * @returns {Promise<*>}
 */
export async function getCallRailInformation({ callrailId }) {
  checkForIdExistence(callrailId);
  let company;
  try {
    const data = await makeCallRailRequest({
      method: 'GET',
      url: `${CALL_RAIL_URL}/companies/${callrailId}.json`,
    });

    company = data.data;
  } catch (e) {
    errorHandling({
      customMessage: 'Failed to retrieve Callrail information',
      e,
    });
  }

  if (company.status === 'disabled') {
    throw statusError(statusError.NOT_FOUND, 'The callRail account has been disabled');
  }
  try {
    const activeTrackers = await getActiveTrackers(callrailId);
    const { data: { integrations } } = await makeCallRailRequest({
      method: 'GET',
      url: `${CALL_RAIL_URL}/integrations.json`,
      params: { company_id: callrailId },
    });

    return {
      company,
      trackers: activeTrackers,
      webhooks: integrations,
    };
  } catch (e) {
    errorHandling({
      customMessage: 'Failed to retrieve Callrail information',
      e,
    });
  }
}

/**
 * Disable the CallRail account associated with a practice account.
 * All the features provided by CallRail will be disabled as well.
 * @throws StatusError
 * @param account
 * @returns {Promise<*|void>}
 */
export async function disableCallRailAccount(account) {
  const { callrailId } = account;
  checkForIdExistence(callrailId);

  try {
    await makeCallRailRequest({
      method: 'DELETE',
      url: `${CALL_RAIL_URL}/companies/${callrailId}.json`,
    });

    return account.update({ callrailId: null });
  } catch (e) {
    errorHandling({
      customMessage: 'Failed to delete the callrail account',
      e,
    });
  }
}

/**
 * Create a Company in CallRail for the given account,
 * enable tracker, pre/post call webhooks for the account.
 * @throws StatusError
 * @param account
 * @returns {Promise<*>}
 */
export async function createCallRailAccount(account) {
  const { destinationPhoneNumber, id, name, callrailId } = account;
  if (callrailId) {
    throw statusError(statusError.CONFLICT, 'Account already has the call rail account associated with');
  }
  if (!destinationPhoneNumber) {
    throw statusError(statusError.NOT_FOUND, 'Destination phone number is not provided');
  }

  try {
    const phoneNumber = destinationPhoneNumber.replace(/\s+/g, '');
    const areaCode = destinationPhoneNumber.replace(/\D/g, '').substr(1, 3);
    const company = await makeCallRailRequest({
      method: 'POST',
      url: `${CALL_RAIL_URL}/companies.json`,
      data: { name },
    });
    const createTracker = {
      method: 'POST',
      url: `${CALL_RAIL_URL}/trackers.json`,
      data: {
        name,
        type: 'source',
        company_id: company.data.id,
        call_flow: {
          type: 'basic',
          recording_enabled: true,
          destination_number: phoneNumber,
        },
        tracking_number: {
          area_code: areaCode,
          local: phoneNumber,
        },
        source: { type: 'offline' },
      },
    };
    await makeCallRailRequest(createTracker);

    const createWebhook = {
      method: 'POST',
      url: `${CALL_RAIL_URL}/integrations.json`,
      data: {
        company_id: company.data.id,
        type: 'Webhooks',
        config: {
          post_call_webhook: [`${fullHostUrl}/callrail/${id}/inbound/post-call`],
          pre_call_webhook: [`${fullHostUrl}/callrail/${id}/inbound/pre-call`],
        },
      },
    };
    await makeCallRailRequest(createWebhook);

    return account.update({ callrailId: company.data.id });
  } catch (e) {
    errorHandling({
      customMessage: 'Failed to create the callrail account',
      e,
    });
  }
}

async function getActiveTrackers(callrailId) {
  const { data: { trackers } } = await makeCallRailRequest({
    method: 'GET',
    url: `${CALL_RAIL_URL}/trackers.json`,
    params: { company_id: callrailId },
  });

  return trackers.filter(({ status }) => status === 'active');
}
