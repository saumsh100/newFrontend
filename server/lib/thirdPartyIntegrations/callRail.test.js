import axios from 'axios';
import * as CallRail from './callRail';
import StatusError from '../../util/StatusError';
import { account } from '../../../tests/util/seedTestUsers';

const oldPhoneNumber = '+123456';
const newPhoneNumber = '+654321';
const callrailId = 123456;
const accountWithCallRailId = {
  ...account,
  callrailId,
  update: jest.fn(),
};
const accountWithOutCallrailId = {
  ...account,
  update: jest.fn(),
};
const testErrorMessage = 'error message';
const callRailError = {
  ...Error,
  statusText: testErrorMessage,
};
const statusError = StatusError(0, testErrorMessage);

jest.mock('axios');

describe('CallRail tests', () => {
  beforeAll(async () => {
    jest.spyOn(console, 'error');
    console.error.mockImplementation(() => {});
  });

  describe('CallRail update phone number tests', () => {
    beforeAll(async () => {
      jest.spyOn(CallRail, 'getCallRailInformation');
      CallRail.getCallRailInformation.mockImplementation(() => {
        return true;
      });
    });
    afterEach(async () => { 
      axios.mockClear();
      CallRail.getCallRailInformation.mockClear();
    });
    afterAll(async () => {
      axios.mockReset();
      CallRail.getCallRailInformation.mockRestore();
    });

    test('CallRail update phone number should throw error when the account does not have callRail account associate with', async () => {
      await expect(CallRail.updatePhoneNumber(accountWithOutCallrailId)).rejects.toThrow(StatusError);
      expect(axios).toHaveBeenCalledTimes(0);
    });

    test('CallRail update phone number should throw error when phone number is not provided', async () => {
      await expect(CallRail.updatePhoneNumber(accountWithCallRailId)).rejects.toThrow(StatusError);
      expect(axios).toHaveBeenCalledTimes(0);
    });

    test('CallRail update phone number should do nothing when the no phone number matches the old phone number', async () => {
      axios.mockReturnValueOnce({ data: { trackers: [{ status: 'active', destination_number: oldPhoneNumber }] } });
      await CallRail.updatePhoneNumber(accountWithCallRailId, newPhoneNumber, newPhoneNumber);
      expect(axios).toHaveBeenCalledTimes(1);
    });

    test('CallRail update phone number should do nothing when the no phone number matches the old phone number', async () => {
      axios.mockReturnValueOnce({ data: { trackers: [{ status: 'active', destination_number: oldPhoneNumber }] } })
      await CallRail.updatePhoneNumber(accountWithCallRailId, oldPhoneNumber, newPhoneNumber);
      expect(axios).toHaveBeenCalledTimes(2);
    });
  });

  describe('CallRail update company name tests', () => {
    afterEach(async () => {
      axios.mockClear();
    });
    afterAll(async () => axios.mockReset());

    test('CallRail update company name should throw error when the account does not have callRail account associate with', async () => {
      await expect(CallRail.updateCompanyName(accountWithOutCallrailId)).rejects.toThrow(StatusError);
      expect(axios).toHaveBeenCalledTimes(0);
    });

    test('CallRail update company name should throw error when the CallRail API failed', async () => {
      axios.mockImplementation(() => Promise.reject(callRailError));
      await expect(CallRail.updateCompanyName(accountWithCallRailId)).rejects.toEqual(statusError);
    });

    test('CallRail update company name should call the CallRail API and success', async () => {
      axios.mockReturnValue({});
      await CallRail.updateCompanyName(accountWithCallRailId);
      expect(axios).toHaveBeenCalledTimes(1);
    });
  });

  describe('CallRail get account tests', () => {
    afterEach(async () => {
      axios.mockClear();
    });
    afterAll(async () => axios.mockReset());

    test('CallRail get account should throw error when the account does not have callRail account associate with', async () => {
      await expect(CallRail.getCallRailInformation(accountWithOutCallrailId)).rejects.toThrow(StatusError);
      expect(axios).toHaveBeenCalledTimes(0);
    });

    test('CallRail get account should throw error when the CallRail API failed', async () => {
      axios.mockImplementation(() => Promise.reject(callRailError));
      await expect(CallRail.getCallRailInformation(accountWithCallRailId)).rejects.toEqual(statusError);
    });

    test('CallRail get account should throw error when the account is disabled', async () => {
      axios.mockReturnValue({ data: { status: 'disabled' } });
      await expect(CallRail.getCallRailInformation(accountWithCallRailId)).rejects.toThrow(StatusError);
    });

    test('CallRail get account should get account information with only active information when success', async () => {
      axios.mockReturnValueOnce({ data: { status: 'active' } })
        .mockReturnValueOnce({ data: { trackers: [{ status: 'active' }, { status: 'disabled' }] } })
        .mockReturnValueOnce({ data: { integrations: [] } });
      await expect(CallRail.getCallRailInformation(accountWithCallRailId)).toMatchSnapshot();
    });
  });

  describe('CallRail delete account tests', () => {
    afterEach(async () => {
      axios.mockClear();
      accountWithCallRailId.update.mockReset();
    });
    afterAll(async () => axios.mockReset());

    test('CallRail delete account should throw error when the account does not have callRail account associate with', async () => {
      await expect(CallRail.disableCallRailAccount(accountWithOutCallrailId)).rejects.toThrow(StatusError);
      expect(axios).toHaveBeenCalledTimes(0);
    });

    test('CallRail delete account should throw error when CallRail API failed and not change the account model', async () => {
      axios.mockImplementation(() => Promise.reject(callRailError));
      await expect(CallRail.disableCallRailAccount(accountWithCallRailId)).rejects.toEqual(statusError);
    });

    test('CallRail delete account should update the account model when success', async () => {
      axios.mockReturnValue({});
      await CallRail.disableCallRailAccount(accountWithCallRailId);
      expect(axios).toHaveBeenCalledTimes(1);
      expect(accountWithCallRailId.update).toHaveBeenCalledTimes(1);
      expect(accountWithCallRailId.update).toBeCalledWith({ callrailId: null });
    });
  });

  describe('CallRail create account tests', () => {
    afterEach(async () => {
      axios.mockClear();
      accountWithOutCallrailId.update.mockReset();
    });
    afterAll(async () => axios.mockReset());

    test('CallRail creates account should throw error when the account already has callRail account associate with', async () => {
      await expect(CallRail.createCallRailAccount(accountWithCallRailId)).rejects.toThrow(StatusError);
      expect(axios).toHaveBeenCalledTimes(0);
    });

    test('CallRail creates account should throw error when CallRail API failed and not change the account model', async () => {
      axios.mockImplementation(() => Promise.reject(callRailError));
      await expect(CallRail.createCallRailAccount(accountWithOutCallrailId)).rejects.toEqual(statusError);
    });

    test('CallRail creates account should update the account model when success', async () => {
      axios.mockReturnValue({ data: { id: callrailId } });
      await CallRail.createCallRailAccount(accountWithOutCallrailId);
      expect(axios).toHaveBeenCalledTimes(3);
      expect(accountWithOutCallrailId.update).toHaveBeenCalledTimes(1);
      expect(accountWithOutCallrailId.update).toBeCalledWith({ callrailId });
    });
  });
});
