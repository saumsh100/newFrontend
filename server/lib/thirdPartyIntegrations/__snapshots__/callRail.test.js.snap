// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallRail tests CallRail get account tests CallRail get account should get account information with only active information when success 1`] = `Promise {}`;
